flink-runtime/src/test/java/org/apache/flink/runtime/resourcemanager/slotmanager/DeclarativeSlotManagerTest.java

```diff
@@ -0,0 +1,1092 @@
+
+			assertThat(slotManager.getNumberRegisteredSlots(), is(1));
+			assertThat(getTotalResourceCount(resourceTracker.getAcquiredResources(jobID)), is(1));
+		}
+	}
+
+	/**
+	 * Tests notify the job manager of the allocations when the task manager is failed/killed.
+	 */
+	@Test
+	public void testNotifyFailedAllocationWhenTaskManagerTerminated() throws Exception {
+		final ResourceTracker resourceTracker = new DefaultResourceTracker();
+
+		try (final DeclarativeSlotManager slotManager = createDeclarativeSlotManagerBuilder()
+			.setResourceTracker(resourceTracker)
+			.buildAndStartWithDirectExec()) {
+
+			// register slot request for job1.
+			JobID jobId1 = new JobID();
+			slotManager.processResourceRequirements(createResourceRequirements(jobId1, 2));
+
+			// create task-manager-1 with 2 slots.
+			final ResourceID taskExecutorResourceId1 = ResourceID.generate();
+			final TestingTaskExecutorGateway testingTaskExecutorGateway1 = new TestingTaskExecutorGatewayBuilder().createTestingTaskExecutorGateway();
+			final TaskExecutorConnection taskExecutionConnection1 = new TaskExecutorConnection(taskExecutorResourceId1, testingTaskExecutorGateway1);
+			final SlotReport slotReport1 = createSlotReport(taskExecutorResourceId1, 2);
+
+			// register the task-manager-1 to the slot manager, this will trigger the slot allocation for job1.
+			slotManager.registerTaskManager(taskExecutionConnection1, slotReport1);
+
+			// register slot request for job2.
+			JobID jobId2 = new JobID();
+			slotManager.processResourceRequirements(createResourceRequirements(jobId2, 2));
+
+			// register slot request for job3.
+			JobID jobId3 = new JobID();
+			slotManager.processResourceRequirements(createResourceRequirements(jobId3, 1));
+
+			// create task-manager-2 with 3 slots.
+			final ResourceID taskExecutorResourceId2 = ResourceID.generate();
+			final TestingTaskExecutorGateway testingTaskExecutorGateway2 = new TestingTaskExecutorGatewayBuilder().createTestingTaskExecutorGateway();
+			final TaskExecutorConnection taskExecutionConnection2 = new TaskExecutorConnection(taskExecutorResourceId2, testingTaskExecutorGateway2);
+			final SlotReport slotReport2 = createSlotReport(taskExecutorResourceId2, 3);
+
+			// register the task-manager-2 to the slot manager, this will trigger the slot allocation for job2 and job3.
+			slotManager.registerTaskManager(taskExecutionConnection2, slotReport2);
+
+			// validate for job1.
+			slotManager.unregisterTaskManager(taskExecutionConnection1.getInstanceID(), TEST_EXCEPTION);
+
+			assertThat(getTotalResourceCount(resourceTracker.getMissingResources().get(jobId1)), is(2));
+
+			// validate the result for job2 and job3.
+			slotManager.unregisterTaskManager(taskExecutionConnection2.getInstanceID(), TEST_EXCEPTION);
+
+			assertThat(getTotalResourceCount(resourceTracker.getMissingResources().get(jobId2)), is(2));
+			assertThat(getTotalResourceCount(resourceTracker.getMissingResources().get(jobId3)), is(1));
```

Comment:
Not sure whether this test needs to be so complex. I guess it would be good enough to test it only with a single TM.